lib/Doctrine/MongoDB/Aggregation/Stage/Operator.php

```diff
@@ -0,0 +1,988 @@
+<?php
+/*
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license. For more information, see
+ * <http://www.doctrine-project.org>.
+ */
+
+namespace Doctrine\MongoDB\Aggregation\Stage;
+
+use Doctrine\MongoDB\Aggregation\Builder;
+use Doctrine\MongoDB\Aggregation\Expr;
+use Doctrine\MongoDB\Aggregation\Stage;
+
+/**
+ * Fluent interface for adding operators to aggregation stages.
+ *
+ * @author alcaeus <alcaeus@alcaeus.org>
+ */
+abstract class Operator extends Stage
+{
+    /**
+     * @var Expr
+     */
+    protected $expr;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function __construct(Builder $builder)
+    {
+        $this->expr = new Expr();
+
+        parent::__construct($builder);
+    }
+
+    /**
+     * Adds numbers together or adds numbers and a date. If one of the arguments is a date, $add treats the other arguments as milliseconds to add to the date.
+     *
+     * The arguments can be any valid expression as long as they resolve to either all numbers or to numbers and a date.
+     *
+     * @see http://docs.mongodb.org/manual/reference/operator/aggregation/add/
+     * @see Expr::add
+     * @param mixed|Expr $expression1
+     * @param mixed|Expr $expression2
+     * @param mixed|Expr $expression3,... Additional expressions
+     * @return self
+     */
+    public function add($expression1, $expression2 /* , $expression3, ... */)
```

Comment:
I mainly wrote this code this way because that's how it was done in ```Query\Builder``` and ```Query\Expr```. To be honest it's painstaking to write, it's a bunch of code duplication and we should actually have an interface that defines those methods to make the classes compatible (in the end it doesn't matter if I call ```add()``` on an ```Expr``` object or the ```Builder``` object itself.
As for code completion, this could be fixed by adding ```@method``` annotations to the class docblock. I know that PHPStorm parses them and offers code completion, don't know about the other IDEs though.